clients/client-py/taskcluster/aio/upload.py

```diff
@@ -0,0 +1,133 @@
+    async def readerFactory():
+        file.seek(0)
+        return FileReader(file)
+
+    await upload(**kwargs, readerFactory=readerFactory)
+
+
+async def upload(*, projectId, name, contentType, contentLength, expires,
+                 readerFactory, maxRetries=5, objectService):
+    """
+    Upload the given data to the object service with the given metadata.
+    The `maxRetries` parameter has the same meaning as for service clients.
+    The `objectService` paraemter is an instance of the Object class,
+    configured with credentials for the upload.
+    """
+    async with aiohttp.ClientSession() as session:
+        uploadId = taskcluster.slugid.nice()
+        proposedUploadMethods = {}
+
+        if contentLength < DATA_INLINE_MAX_SIZE:
+            reader = await readerFactory()
+            writer = BufferWriter()
+            await streamingCopy(reader, writer)
+            encoded = base64.b64encode(writer.getbuffer())
+            proposedUploadMethods['dataInline'] = {
+                "contentType": contentType,
+                "objectData": encoded,
+            }
+
+        proposedUploadMethods['putUrl'] = {
+            "contentType": contentType,
+            "contentLength": contentLength,
+        }
+
+        uploadResp = await objectService.createUpload(name, {
+            "expires": expires,
+            "projectId": projectId,
+            "uploadId": uploadId,
+            "proposedUploadMethods": proposedUploadMethods,
+        })
+
+        async def tryUpload(retryFor):
+            try:
+                uploadMethod = uploadResp["uploadMethod"]
+                if 'dataInline' in uploadMethod:
+                    # data is already uploaded -- nothing to do
+                    pass
+                elif 'putUrl' in uploadMethod:
+                    reader = await readerFactory()
+                    await _putUrlUpload(uploadMethod['putUrl'], reader, session)
+                else:
+                    raise RuntimeError("Could not negotiate an upload method")
+            except aiohttp.ClientResponseError as exc:
+                # treat 4xx's as fatal, and retry others
+                if 400 <= exc.status < 500:
+                    raise exc
+                return retryFor(exc)
+            except aiohttp.ClientError as exc:
+                # retry for all other aiohttp errors
+                return retryFor(exc)
+            # .. anything else is considered fatal
+
+        await retry(maxRetries, tryUpload)
+
+        await objectService.finishUpload(name, {
+            "projectId": projectId,
+            "uploadId": uploadId,
+        })
+
+
+async def _putUrlUpload(method, reader, session):
+    chunk_size = 64 * 1024
+
+    async def reader_gen():
+        while True:
+            chunk = await reader.read(chunk_size)
+            if not chunk:
+                break
+            yield chunk
+
+    resp = await session.put(method['url'], headers=method['headers'], data=reader_gen())
```

Comment:
If there is an error, does the caller have a way of logging the response body?

Reply:
I believe the aiohttp exception raised in that case does contain the response, from which the body can be read.

Reply:
oh cool!